packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart

```diff
@@ -0,0 +1,92 @@
+import 'package:amplify_core/amplify_core.dart';
+import 'package:amplify_push_notifications/amplify_push_notifications.dart';
+import 'package:amplify_push_notifications/src/push_notifications_flutter_api.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:mockito/mockito.dart';
+import 'package:os_detect/override.dart';
+import 'package:shared_preferences/shared_preferences.dart';
+
+import 'amplify_push_noitfications_impl_test.mocks.dart';
+import 'test_data/fake_notification_messges.dart';
+
+void testGlobalCallbackFunction(PushNotificationMessage pushMessage) {
+  expect(
+    pushMessage.title,
+    PushNotificationMessage.fromJson(standardAndroidPushMessage).title,
+  );
+}
+
+void main() {
+  late AmplifyPushNotificationsFlutterApi flutterApi;
+
+  setUp(() {
+    TestWidgetsFlutterBinding.ensureInitialized();
+    flutterApi = AmplifyPushNotificationsFlutterApi.instance;
+  });
+
+  test('should queue events when the serviceClient is not ready', () {
+    SharedPreferences.setMockInitialValues({});
+    flutterApi.onNotificationReceivedInBackground(standardAndroidPushMessage);
+    expect(flutterApi.eventQueue.length, 1);
+  });
+
+  test('should flush events when the serviceClient is ready', () async {
+    SharedPreferences.setMockInitialValues({});
+    await flutterApi
+        .onNotificationReceivedInBackground(standardAndroidPushMessage);
+    expect(flutterApi.eventQueue.length, 2);
+    final mockServiceClient = MockServiceProviderClient();
+    when(
+      mockServiceClient.recordNotificationEvent(
+        eventType: anyNamed('eventType'),
+        notification: anyNamed('notification'),
+      ),
+    ).thenAnswer(
+      (_) async {},
+    );
+    flutterApi.serviceProviderClient = mockServiceClient;
+    await Future.delayed(const Duration(seconds: 3), () {});
+    expect(flutterApi.eventQueue.length, 0);
+  });
+
+  test('should call the external callback function', () async {
+    SharedPreferences.setMockInitialValues({});
+    void externalCallback(PushNotificationMessage pushMessage) {
+      expect(
+        pushMessage.title,
+        PushNotificationMessage.fromJson(standardAndroidPushMessage).title,
+      );
+    }
+
+    flutterApi.registerOnReceivedInBackgroundCallback(externalCallback);
+    await flutterApi
+        .onNotificationReceivedInBackground(standardAndroidPushMessage);
+  });
+
+  test('should invoke the top-level or static external callback function',
+      () async {
+    await overrideOperatingSystem(
+      const OperatingSystem('android', ''),
+      () async {
+        SharedPreferences.setMockInitialValues({});
+        final pref = await SharedPreferences.getInstance();
+        AmplifyPushNotifications(
+          serviceProviderClient: MockServiceProviderClient(),
+          backgroundProcessor: () async => {},
+          // dependencyManager: dependencyManager,
+        ).onNotificationReceivedInBackground(
+          testGlobalCallbackFunction,
```

Comment:
```suggestion
          expectAsync1(testGlobalCallbackFunction),
```

You need to set an expectation that this will be called. As it stands, this test would complete successfully whether the function is called or not.

Reply:
Tried `expectAsync1` but won't work for this case because `onNotificationReceivedInBackground` expects the callback function provided to be global or static but when I use `expectAsync1`, I believe it wraps it and makes it a local function?

Reply:
There should be no state left over between tests - this will greatly increase the risk of false positives in tests (like here potentially). I would recommend adding a `reset` method to the static class and call that in between tests.

Reply:
You can implement it yourself with completers